eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php

```diff
@@ -1891,4 +1893,151 @@ public function copyRelations($originalContentId, $copiedContentId, $versionNo =
+     * Remove the specified translation from the Content Object Version.
+     *
+     * @param int $contentId
+     * @param string $languageCode language code of the translation
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function removeTranslationFromContent($contentId, $languageCode)
+    {
+        $language = $this->languageHandler->loadByLanguageCode($languageCode);
+        $this->connection->beginTransaction();
+        try {
+            $this->deleteTranslationFromContentVersions($contentId, $language->id);
+            $this->deleteTranslationFromContentObject($contentId, $language->id);
+
+            $this->deleteTranslationFromContentAttributes($contentId, $languageCode);
+            $this->deleteTranslationFromContentNames($contentId, $languageCode);
+
+            $this->connection->commit();
+        } catch (DBALException $e) {
+            $this->connection->rollBack();
+            throw $e;
+        }
+    }
+
+    /**
+     * Delete translation from the ezcontentobject_attribute table.
+     *
+     * @param int $contentId
+     * @param string $languageCode
+     */
+    private function deleteTranslationFromContentAttributes($contentId, $languageCode)
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query
+            ->delete('ezcontentobject_attribute')
+            ->where('contentobject_id = :contentId')
+            ->andWhere('language_code = :languageCode')
+            ->setParameters(
+                [
+                    ':contentId' => $contentId,
+                    ':languageCode' => $languageCode,
+                ]
+            )
+        ;
+
+        $query->execute();
+    }
+
+    /**
+     * Delete translation from the ezcontentobject_name table.
+     *
+     * @param $contentId
+     * @param $languageCode
+     */
+    private function deleteTranslationFromContentNames($contentId, $languageCode)
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query
+            ->delete('ezcontentobject_name')
+            ->where('contentobject_id=:contentId')
+            ->andWhere('real_translation=:languageCode')
+            ->setParameters(
+                [
+                    ':languageCode' => $languageCode,
+                    ':contentId' => $contentId,
+                ]
+            )
+        ;
+
+        $query->execute();
+    }
+
+    /**
+     * Remove language from language_mask of ezcontentobject.
+     *
+     * @param int $contentId
+     * @param int $languageId
+     * @throws \eZ\Publish\Core\Base\Exceptions\BadStateException
+     */
+    private function deleteTranslationFromContentObject($contentId, $languageId)
+    {
+        $query = $this->connection->createQueryBuilder();
+        $query->update('ezcontentobject')
+            // parameter for bitwise operation has to be placed verbatim (w/o binding) for this to work cross-DBMS
+            ->set('language_mask', 'language_mask & ~ ' . $languageId)
```

Comment:
Note: A little bit of bitwise voodoo that seems to be working for MySQL, Sqlite and PostgreSQL (thx @Nattfarinn :) ). 
I didn't bind it as a parameter because PostgreSQL cannot determine data type then and demands explicit cast... which is DBMS-specific.
Anyway, `$languageId` comes from `$this->languageHandler->loadByLanguageCode($languageCode);` so we can safely assume it's integer.